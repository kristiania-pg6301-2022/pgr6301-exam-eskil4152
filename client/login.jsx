import { useNavigate } from "react-router-dom";
import { useEffect, useState } from "react";

export function Register() {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [fullName, setFullName] = useState("");
  const [newUser, setNewUser] = useState({ username, password, fullName });
  const [error, setError] = useState("");

  useEffect(() => {
    setNewUser({ username, password, fullName });
  }, [username, password, fullName]);

  const navigate = useNavigate();

  async function handleSubmit(event) {
    event.preventDefault();
    const res = await fetch("/api/login/new", {
      method: "post",
      body: JSON.stringify({ username, password, fullName }),
      headers: {
        "content-type": "application/json",
      },
    });

    if (res.ok) {
      navigate("..");
    } else {
      setError("Error: " + error.toString());
    }
  }

  return (
    <form onSubmit={handleSubmit}>
      <div>
        Username:{" "}
        <input
          value={username}
          name={"username"}
          onChange={(e) => setUsername(e.target.value)}
        />
      </div>
      <div>
        Full name:{" "}
        <input
          value={fullName}
          name={"fullName"}
          onChange={(e) => setFullName(e.target.value)}
        />
      </div>
      <div>
        Password:{" "}
        <input
          value={password}
          type={"password"}
          name={""}
          onChange={(e) => setPassword(e.target.value)}
        />
      </div>
      <div>
        <button>Register</button>
      </div>
    </form>
  );
}

export function Login() {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState("");

  const navigate = useNavigate();

  async function handleSubmit(e) {
    e.preventDefault();
    const res = await fetch("/api/login", {
      method: "post",
      body: JSON.stringify({ username, password }),
      headers: {
        "content-type": "application/json",
      },
    });

    if (res.ok) {
      navigate("..");
    } else {
      setError("Username or password is wrong");
    }
  }

  return (
    <form onSubmit={handleSubmit}>
      <h1>Please log in</h1>
      <div>
        Username:{" "}
        <input value={username} onChange={(e) => setUsername(e.target.value)} />
      </div>
      <div>
        Password:{" "}
        <input
          type="password"
          value={password}
          onChange={(e) => setPassword(e.target.value)}
        />
      </div>
      <div>{error}</div>
      <div>
        <button>Log in</button>
      </div>
    </form>
  );
}
