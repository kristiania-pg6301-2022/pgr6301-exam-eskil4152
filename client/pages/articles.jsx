import React, { useContext, useEffect, useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import { fetchJSON } from "../http";
import { useLoading } from "../useLoading";

{
  /*var CATEGORIES = ["Sports", "Breaking", "Drama", "Technology", "Economy", "World"]
    const sc = document.getElementById("selectCategory");

    for(var i = 0; i < CATEGORIES.length; i++) {
        var opt = CATEGORIES[i]
        var el = document.createElement("option")
        el.textContent = opt
        el.value = opt
        sc.append(el)
    }*/
}

function ArticleCard({ article }) {
  const { title, category, text, author } = article;
  return (
    <div>
      <h3>
        {title}, {category}
      </h3>
      <h5>Author: {author}</h5>
      <p>{text}</p>
    </div>
  );
}

function ArticlePreviewCard({ article }) {
  const { title, category } = article;
  return (
    <div>
      <h4>
        {title}, {category}
      </h4>
    </div>
  );
}

const Context = React.createContext({
  async listArticles() {
    return await fetchJSON("/api/articles");
  },
});

export function FullArticles() {
  const { listArticles } = useContext(Context);

  const { loading, error, data } = useLoading(async () => await listArticles());

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  return (
    <div>
      <h1>Articles</h1>
      <a href={"/articles/new"}>
        <button>New article</button>
      </a>
      <a href={"/articles/edit"}>
        <button>Edit article</button>
      </a>
      {data.map((article) => (
        <div>
          <ArticleCard key={article.title} article={article} />
        </div>
      ))}
    </div>
  );
}

export function EditArticle() {
  const navigate = useNavigate();

  const [title, setTitle] = useState("");
  const [newTitle, setNewTitle] = useState("");
  const [newCategory, setNewCategory] = useState("");
  const [newText, setNewText] = useState("");
  const [error, setError] = useState("");

  const [editedArticle, setEditedArticle] = useState({
    title,
    newTitle,
    newCategory,
    newText,
  });

  useEffect(() => {
    setEditedArticle({ title, newTitle, newCategory, newText });
  }, [title, newTitle, newCategory, newText]);

  async function handleSubmit(event) {
    event.preventDefault();
    const res = await fetch("/api/articles", {
      method: "put",
      body: JSON.stringify({ title, newTitle, newCategory, newText }),
      headers: {
        "content-type": "application/json",
      },
    });

    if (res.ok) {
      navigate("..");
    } else {
      setError("You did not make this article");
    }
  }

  return (
    <form onSubmit={handleSubmit}>
      <div>
        Name of article to edit:{" "}
        <input
          type={"text"}
          name={"title"}
          onChange={(e) => setTitle(e.target.value)}
        />
      </div>
      <div>
        New name:{" "}
        <input
          type={"text"}
          name={"newTitle"}
          onChange={(e) => setNewTitle(e.target.value)}
        />
      </div>
      <div>
        New category:{" "}
        <input
          type={"text"}
          name={"newCategory"}
          onChange={(e) => setNewCategory(e.target.value)}
        />
      </div>
      <div>
        New text:
        <div>
          <textarea
            name={"newText"}
            onChange={(e) => setNewText(e.target.value)}
          />
        </div>
      </div>
      <div>{error}</div>
      <button>Submit</button>
    </form>
  );
}

export function ArticlesPreview() {
  const { listArticles } = useContext(Context);

  const { loading, error, data } = useLoading(async () => await listArticles());

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  return (
    <div>
      <h1>Articles</h1>
      {data.map((article) => (
        <ArticlePreviewCard key={article.title} article={article} />
      ))}
    </div>
  );
}

export function NewArticle() {
  const [title, setTitle] = useState("");
  const [category, setCategory] = useState("");
  const [text, setText] = useState("");
  const [newArticle, setNewArticle] = useState();
  const navigate = useNavigate();

  useEffect(() => {
    setNewArticle({ title, category, text });
  }, [title, category, text]);

  async function handleSubmit(event) {
    event.preventDefault();
    const res = await fetch("/api/articles", {
      method: "post",
      body: JSON.stringify({ title, category, text }),
      headers: {
        "content-type": "application/json",
      },
    });
    navigate("..");
  }

  return (
    <form onSubmit={handleSubmit}>
      <div>
        Title:{" "}
        <input
          value={title}
          name={"title"}
          onChange={(e) => setTitle(e.target.value)}
        />
      </div>
      <div>
        Category:{" "}
        <input
          value={category}
          name={"category"}
          onChange={(e) => setCategory(e.target.value)}
        />
        {/*<select value={category} name={"category"} id={"selectCategory"} onChange={(e) => setCategory(e.target.value)}><option>Select</option></select>*/}
      </div>
      <div>
        Text:{" "}
        <div>
          <textarea
            value={text}
            name={"text"}
            onChange={(e) => setText(e.target.value)}
          />
        </div>
      </div>
      <div>
        <button>Submit</button>
      </div>
    </form>
  );
}
