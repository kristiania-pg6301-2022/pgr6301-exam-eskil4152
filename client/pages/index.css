#app {
    position: absolute;
    height: 100%;
    width: 100%;
    display: grid;
    grid-template-rows: auto 1fr;
}

main {
    position: absolute;
    overflow-y: auto;
    margin-left: 150px;
    padding-top: 35px;
    padding-left: 20px;
    height: auto;
}

#cards {
    background-color: whitesmoke;
}

sidebar {
    background-color: lightgrey;
    width: 150px;
    padding-left: 10px;
    border: 1px solid black;
}

body {
    margin: 0;
}

header {
    background-color: lightgrey;
    padding: 0.5em 1em;
    display: flex;
    justify-content: space-between;
    border: 1px solid black;
}

header > * {
    padding-inline: 0.5em;
}

header .menu-divider {
    flex: 1;
}
