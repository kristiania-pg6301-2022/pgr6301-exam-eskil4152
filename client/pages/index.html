<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Index</title>
</head>
<body>
<div id="app"></div>
</body>
<script src="../application.jsx" type="module"></script>
</html>