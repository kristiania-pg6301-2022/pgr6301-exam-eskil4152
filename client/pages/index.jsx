import ReactDOM from "react-dom";
import React from "react";
import {BrowserRouter, Link, Route, Routes, useNavigate,} from "react-router-dom";
import {ArticlesPreview, EditArticle, FullArticles, NewArticle} from "./articles";
import {useLoader} from "../useLoader";
import {fetchJSON} from "../http";
import "./index.css"
import {Login, Register, Logout} from "./login";

function Application() {
  return (
    <BrowserRouter>
        <header>
            <Link to={"/"}>Front page</Link>
            <div className="menu-divider" />
            <Link to={"/logout"}>Log out</Link>
        </header>
        <sidebar>
            <div><ArticlesPreview /></div>
        </sidebar>
        <main>
      <Routes>
          <Route path={"/"} element={<FrontPage />} />
          <Route path={"/login"} element={<Login />} />
          <Route path={"/register"} element={<Register />} />
          <Route path={"/articles/new"} element={<NewArticle />} />
          <Route path={"/articles/edit"} element={<EditArticle />} />
          <Route path={"/logout"} element={<Logout />} />
          <Route path={"*"} element={<h1>Not found</h1>} />
      </Routes>
        </main>
    </BrowserRouter>
  );
}

function FrontPage() {
    const navigate = useNavigate();
  const { loading, error, data, reload } = useLoader(
    async () => await fetchJSON("/api/login")
  );
  const user = data;

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div style={{ border: "1px solid red", background: "Pink" }}>
        An error occurred: {error.toString()}
      </div>
    );
  }

    return <div>{!user ? <LoginRegister /> : <FullArticles />}</div>
}

function LoginRegister() {
  return (
    <div>
      <h1>Welcome to my webpage</h1>
        <h3>You have to log in to see full articles</h3>
      <div>
        <a href={"/login"}>Log in</a>
      </div>
      <div>
        <a href={"/register"}>Register</a>
      </div>
    </div>
  );
}

ReactDOM.render(<Application />, document.getElementById("app"));
