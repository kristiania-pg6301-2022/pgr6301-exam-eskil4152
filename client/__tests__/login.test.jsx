import { Login, Logout, Register } from "../pages/login";
import ReactDOM from "react-dom";
import React from "react";
import { MemoryRouter } from "react-router-dom";

describe("login tests", () => {
  it("should return login page", function () {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <Login />
      </MemoryRouter>,
      element
    );

    expect(element).toMatchSnapshot();
  });

  it("should return register page", function () {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <Register />
      </MemoryRouter>,
      element
    );

    expect(element).toMatchSnapshot();
  });

  it("should return logout page", function () {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <Logout />
      </MemoryRouter>,
      element
    );

    expect(element).toMatchSnapshot();
  });
});
