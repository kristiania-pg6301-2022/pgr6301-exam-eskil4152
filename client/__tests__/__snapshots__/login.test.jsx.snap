// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login tests should return login page 1`] = `
<div>
  <form>
    <h1>
      Please log in
    </h1>
    <div>
      Username:
       
      <input
        value=""
      />
    </div>
    <div>
      Password:
       
      <input
        type="password"
        value=""
      />
    </div>
    <div />
    <div>
      <button>
        Log in
      </button>
    </div>
  </form>
</div>
`;

exports[`login tests should return logout page 1`] = `
<div>
  <form>
    <div>
      <h1>
        Are you sure?
      </h1>
      <button>
        Log out
      </button>
    </div>
  </form>
</div>
`;

exports[`login tests should return register page 1`] = `
<div>
  <form>
    <div>
      Username:
       
      <input
        name="username"
        value=""
      />
    </div>
    <div>
      Full name:
       
      <input
        name="fullName"
        value=""
      />
    </div>
    <div>
      Password:
       
      <input
        name=""
        type="password"
        value=""
      />
    </div>
    <div>
      <button>
        Register
      </button>
    </div>
  </form>
</div>
`;
